tests/test_migrations.py

```diff
@@ -0,0 +1,268 @@
+# Necessary to be loaded into the PYTHONPATH so that migrations referencing it can be loaded
+utils.load_module("migration_steps", files_dir / "migration_steps.py")
+migrations = utils.load_module("migrations", files_dir / "migrations.py")
+
+
+INSTALL = 1
+UPGRADE = 2
+VERSION_TARGET = migrations.Version("1.1.0")
+
+
+def _is_missing(path):
+    if path.exists():
+        assert False
+
+
+def test_version_double_digits():
+    assert migrations.Version("0.0.12").version[2] == 12
+
+
+def test_version_ignore_intermediary_releases():
+    assert migrations.Version("20.0.1-rc5").version == migrations.Version("20.0.1").version
+
+
+def test_success_install(tmp_success):
+    version_file = tmp_success / "version"
+    version_file.unlink()
+    _is_missing(version_file)
+
+    migrations.main(version_file, tmp_success, INSTALL, VERSION_TARGET)
+
+    assert version_file.open().read() == str(VERSION_TARGET)
+
+
+def test_success_install_despite_failing_migrations(tmp_failure):
+    version_file = tmp_failure / "version"
+    version_file.unlink()
+    _is_missing(version_file)
+
+    migrations.main(version_file, tmp_failure, INSTALL, VERSION_TARGET)
+
+    assert version_file.open().read() == str(VERSION_TARGET)
+
+
+def test_success_migrate(tmp_success):
+    version_file = tmp_success / "version"
+    assert version_file.open().read() == "0.0.0"
+
+    migrations.main(version_file, tmp_success, UPGRADE, VERSION_TARGET)
+
+    example_file = tmp_success / "example-file.txt"
+    assert example_file.open().read() == "0.0.1"
+
+    assert version_file.open().read() == str(VERSION_TARGET)
+
+
+def test_success_migrate_multiple_and_order(tmp_success):
+    version_file = tmp_success / "version"
+    assert version_file.open().read() == "0.0.0"
+
+    example_file = tmp_success / "example-file.txt"
+
+    for ver, new_migration in [(f"0.0.{i}", tmp_success / f"0.0.{i}.py") for i in range(1, 5)]:
+        if not new_migration.exists():
+            shutil.copy(tmp_success / "0.0.1.py", new_migration)
+
+        migrations.main(version_file, tmp_success, UPGRADE, migrations.Version(ver))
+
+        example_file = tmp_success / "example-file.txt"
+        assert example_file.open().read() == ver
+
+        assert version_file.open().read() == ver
+
+
+def test_validation_failed_partial_success(tmp_success_if_valid, caplog):
+    version_file = tmp_success_if_valid / "version"
+    assert version_file.open().read() == "0.0.0"
+
+    # migrations/success-if-valid.py (which is copied to 0.0.1.py) uses the filename's version as a
+    # base for its operations, and requires sequential execution. The patch version - 1 having been
+    # run before is checked during validation
+    shutil.copy(tmp_success_if_valid / "0.0.1.py", tmp_success_if_valid / "0.0.3.py")
+
+    with pytest.raises(SystemExit) as exited_e:
+        migrations.main(version_file, tmp_success_if_valid, UPGRADE, VERSION_TARGET)
+
+    assert exited_e.type == SystemExit
```

Comment:
I think you don't need this assertion because `pytest.raises` only catches SystemExits?